composer/callbacks/mlperf.py

```diff
@@ -0,0 +1,345 @@
+        target (float, optional): The target metric before the mllogger marks the stop
+            of the timing run. Default: ``0.759`` (resnet benchmark).
+        division (str, optional): Division of submission. Currently only ``open`` division supported.
+        metric_name (str, optional): name of the metric to compare against the target. Default: ``Accuracy``.
+        metric_label (str, optional): label name. The metric will be accessed via ``state.current_metrics[metric_label][metric_name]``.
+        submitter (str, optional): Submitting organization. Default: MosaicML.
+        system_name (str, optional): Name of the system (e.g. 8xA100_composer). If
+            not provided, system name will default to ``[world_size]x[device_name]_composer``,
+            e.g. ``8xNVIDIA_A100_80GB_composer``.
+        status (str, optional): Submission status. One of (onprem, cloud, or preview).
+            Default: ``"onprem"``.
+        cache_clear_cmd (str, optional): Command to invoke during the cache clear. This callback
+            will call ``os.system(cache_clear_cmd)``. Default is disabled (None)
+    """
+
+    def __init__(
+        self,
+        root_folder: str,
+        index: int,
+        benchmark: str = 'resnet',
+        target: float = 0.759,
+        division: str = 'open',
+        metric_name: str = 'Accuracy',
+        metric_label: str = 'eval',
+        submitter: str = "MosaicML",
+        system_name: Optional[str] = None,
+        status: str = "onprem",
+        cache_clear_cmd: Optional[str] = None,
+    ) -> None:
+
+        require_mlperf_logging()
+
+        if benchmark not in BENCHMARKS:
+            raise ValueError(f"benchmark: {benchmark} must be one of {BENCHMARKS}")
+        if division not in DIVISIONS:
+            raise ValueError(f"division: {division} must be one of {DIVISIONS}")
+        if status not in STATUS:
+            raise ValueError(f"status: {status} must be one of {STATUS}")
+
+        self.mllogger = mllog.get_mllogger()
+        self.target = target
+        self.system_name = system_name
+        self.benchmark = benchmark
+        self.root_folder = root_folder
+        self.metric_name = metric_name
+        self.metric_label = metric_label
+
+        system_desc = get_system_description(submitter, division, status, system_name)
+        system_name = system_desc['system_name']
+
+        if dist.get_local_rank() == 0:
+            self._create_submission_folders(root_folder, system_name, benchmark)
+            systems_path = os.path.join(root_folder, 'systems', f'{system_name}.json')
+            with open(systems_path, 'w') as f:
+                json.dump(system_desc, f, indent=4)
+
+        dist.barrier()
+
+        self.filename = os.path.join(root_folder, 'results', system_name, benchmark, f'result_{index}.txt')
+        self.upload_name = '{run_name}' + f'/results/{system_name}/{benchmark}/result_{index}.txt'
```

Comment:
Can this include the node_rank or global_rank? Otherwise, if one is doing multi-node training, the nodes will override each-other.

Reply:
We cannot -- the naming for the mlperf submission package is fairly strict. Upload is only happening on global rank zero so this is not an issue.